streams/src/main/java/org/apache/kafka/streams/state/internals/CachingTimestampedKeyValueStore.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.serialization.LongDeserializer;
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.streams.state.StateSerdes;
+
+import java.nio.ByteBuffer;
+
+class CachingTimestampedKeyValueStore<K, V> extends CachingKeyValueStore<K, V> {
+    private final LongDeserializer longDeserializer = new LongDeserializer();
+
+    CachingTimestampedKeyValueStore(final KeyValueStore<Bytes, byte[]> underlying,
+                                    final Serde<K> keySerde,
+                                    final Serde<V> valueSerde) {
+        super(underlying, keySerde, valueSerde);
+    }
+
+    @Override
+    public <KF, VF> FlushEntry<KF, VF> flushEntry(final StateSerdes<KF, VF> serdes,
+                                                  final byte[] rawKey,
+                                                  final byte[] rawValueAndTimestamp,
+                                                  final byte[] oldRawValueAndTimestamp,
+                                                  final long timestamp) {
+        final VF value;
+        final long valueTimestamp;
+
+        if (rawValueAndTimestamp != null) {
```

Comment:
The logic of serde valueAndTimestamp is scattered in multiple places, like here and in ValueAndTimestampSerde. Could we just reuse the logic in the latter class and not duplicating the logic?

Reply:
I see the desire but it's a little tricky.

Here, we only have `StateSerde<K,V>` and there is no `ValueAndTimestampSerde` instance we could use. What we essentially need here, is to split a `byte[]` that contains value-and-timestamp into the value- and timestamp component. Also we need to deserialize both components.

For `ChangeloggingTimestampedKeyValueStore`, it's slightly different, as we need to split a `byte[]` into both components, too, but need the value-part as `byte[]` and only the timestamp is deserialized.

Thus, what we could do it, to add static methods to `ValueAndTimestampSerde` (or maybe `ValueAndTimestampDeserializer` because it's only about deserialization) to consolidate this logic. Something like:
 - `byte[] valueFromValueAndTimestamp(byte[] rawValueAndTimestamp)`: returns raw value part
 - `long timestampFromValueAndTimestamp(byte[] rawValueAndTimestamp)`: extracts timestamp and returns it as long

WDYT?

Reply:
Yeah, this sounds good to me. We don't have any obligation to stick with the normal `Serde` patterns, and it would be more efficient if we can decompose just the pieces we need when we don't need both. Regardless, it would be nice to consolidate the logic, and static methods on ValueAndTimestampSerde seems like the place to do it.